keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a valid field of type string, boolean or integer. Strings with enum values
+  // and strings with formats are allowed.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+
+  // valueIfAbsent provides a value for field selection if the value referred to
+  // by jsonPath is absent.
+  // valueIfAbsent is required unless all parts of jsonPath are required fields, or have defaults.
+  // +Optional
+  ValueIfAbsent *string `json:"valueIfAbsent" protobuf:"bytes,2,opt,name=valueIfAbsent"`
+}
+```
+
+##### Validaiton rules
+
+- `selectableFields[].jsonPath` must be a "simple path" (similar to
+  `additionalPrinterColumns[].jsonPath` except validated properly). I.e.
+    property names separated by '.', e.g. `.spec.x.y` (The path may not contain
+    map or array indexing)
+- The schema type referenced by `selectableFields[].jsonPath` must be one of:
+  string, integer, boolean (enums and fields with string formats are allowed)
+- If (and only if) the schema type referenced by `selectableFields[].jsonPath`
+  is optional (directly or indirectly), a `selectableFields[].ifAbsent` field
+  must provide a fallback string value.
+- A maximum of 8 `selectableFields` per CRD version are allowed.
+- `selectableFields` may not refer to metadata fields.
+- `selectableFields` may not contain duplicates.
+
+
+#### Discovery
+
+We do not publish any field selector information into OpenAPIv3, and we do not
+document in APIs or on the Kubernetes website which fields may be selected.
+
+Finding the field selectors for built-in types, for the above table, required
+locating [`GetAttrs/*ToSelectableFields` functions in the Kubernetes source code](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L340).
+
+We will add discovery details in to OpenAPIv3:
+
+```json
+"io.k8s.api.core.v1.Pod": {
+        "description": "Pod is a collection of containers that can run on a host. This resource is created by clients and scheduled onto hosts.",
+        "properties": { ... },
+        "type": "object",
+        "x-kubernetes-group-version-kind": [ ... ],
+        "x-kubernetes-selectable-fields": [
+          { "fieldPath": "spec.nodeName"},
+          { "fieldPath": "spec.restartPolicy"},
+          { "fieldPath": "spec.schedulerName"},
+          { "fieldPath": "spec.serviceAccountName"},
+          { "fieldPath": "spec.hostNetwork", "valueIfAbsent": "false", "description": "Set to the value of the spec.securityContext.hostNetwork field"},
+          { "fieldPath": "status.phase"},
+          { "fieldPath": "status.podIP", "valueIfAbsent": "", "description": "Set to the value of status.podIPs[0]"},
+          { "fieldPath": "status.nominatedNodeName"},
+        ]
+      },
+```
+
+#### Declarative field selector definitions on built-in types
```

Comment:
Because field selectors are based on the internal versions of the struct and the same field selector is used regardless of the location in the serialization, I don't understand how you could support this marker on our external types and to my knowledge we lack markers on the internal versions today.

Reply:
That's true... I haven't tried to break ground on this part of the implementation and don't have any good ideas for how I'd try to do this.

I'll mull it over.  I might need to drop built-in declarative work from this KEP if I don't come up with anything.

Reply:
I would really like to see a declarative way of doing that.

Doesn't `AddFieldLabelConversion` function allow us for achieving it:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/events/v1/conversion.go#L63

[I didn't think about it carefully, but I think it should make it possible]

Reply:
I'm going to prototype this.

In the meantime, I'm moving the declarative built-in support to future work.  I'm time constrained on what  I can get done in 1.30, and so I'd like to narrow this KEP just to the CRD support and then circle back next release on built-in improvements and "in" and "not in" support.

Reply:
> so I'd like to narrow this KEP just to the CRD support

I think this is reasonable.  Leave existing as-is, make new use-cases possible.